Adds JSON support

banacorn opened the pull request:
This PR adds a new flag `--interaction-json` for Agda to **output interactions in JSON format**.

----

Emacs was the sole editor for interacting with Agda, but things have changed as people are also trying to interact with Agda with other editors such as [*Atom (agda-mode)*](https://github.com/banacorn/agda-mode) and [*Vim (agda-vim)*](https://github.com/derekelkins/agda-vim).

To interact with Agda, an editor would have to 
1. send a request in the format of some Haskell data value (`IOTCM`) to Agda. (*easy*)
2. receive several responses in the format of **Emacs Lisp** from Agda. (*hard!!!!*)

AFAIK the protocol is not officially documented. It takes quite some effort to [reverse engineer](https://github.com/banacorn/agda-mode/wiki/Conversations-between-Agda-&-agda-mode) and emulate such a protocol, and deciphering those Emacs Lisp messages was the most tricky part.

----

This PR:
* adds a new module `Agda.Interaction.JSONTop` as the counterpart of `Agda.Interaction.EmacsTop` as suggested by @nad in https://github.com/agda/agda/issues/3113#issuecomment-395373758.
* adds `Aeson` as a dependency for JSON encoding support.
* modifies some other files to enable the `--interaction-json` flag.

banacorn committed:
[ JSON ] Dummy module and command line option

banacorn committed:
[ JSON ] add Aeson as dependency

banacorn committed:
[ JSON ] Encode Response in JSON format

banacorn committed:
[ JSON ] Update flag-related golden files in tests

gallais commented on doc/user-manual/tools/command-line-options.rst:
```diff
@@ -27,6 +27,10 @@ General options
       yourself)
 
+:samp:`--interaction`
+    For use with other editors such as Atom (no need to invoke
```
Should this read `interaction-json`?

banacorn replied:
Oopsy!!

gallais commented on src/full/Agda/Interaction/Options.hs:
```diff
@@ -298,6 +300,8 @@ checkOpts :: Flag CommandLineOptions
       throwError "Choose at most one: input file or --interaction.\n"
+  | not (matches [optJSONInteraction, isJust . optInputFile] <= 1) =
+      throwError "Choose at most one: input file or --interaction-json.\n"
   | or [ p opts && matches ps > 1 | (p, ps) <- exclusive ] =
```
I'm guessing it does not make sense to use `interaction` and
`interaction-json` together either. So instead I would write
only one guard for the three possibilities:

```haskell
not (matches [optGHCiInteraction, optJSONInteraction, isJust . optInputFile] <= 1) =
  throwError "Choose at most one: input file, --interactive, or --interaction-json.\n"
```

gallais commented on src/full/Agda/Interaction/Highlighting/JSON.hs:
```diff
@@ -0,0 +1,117 @@
+-- Read/show functions
+
+-- | Converts the 'aspect' and 'otherAspects' fields to atoms
+toAtoms :: Aspects -> [String]
```
This is an exact copy of `toAtoms` from `Agda.Interaction.Highlighting.Emacs`.
You should create `Agda.Interaction.Highlighting.Common` and put such shared
definitions there to avoid code duplication.

banacorn replied:
Will do!

gallais commented on src/full/Agda/Interaction/Highlighting/JSON.hs:
```diff
@@ -0,0 +1,117 @@
+     -- ^ Must contain a mapping for every definition site's module.
+  -> IO Value
+jsonifyHighlightingInfo info remove method modFile = do
+  case ranges info of
```
This shares a lot of the code with `lispifyHighlightingInfo`.
Is there no way to refactor the common parts?

gallais commented on src/full/Agda/Interaction/JSONTop.hs:
```diff
@@ -0,0 +1,203 @@
+
+jsonREPL :: TCM () -> TCM ()
+jsonREPL setup = do
+    liftIO $ do
```
Same here. Looks like we could define an `agdaTop` which would
subsume `jsonREPL` and `mimicGHCi`. It seems to only need to take
as an argument the callback function and the prompt string.

gallais reviewed (commented):
Nice work & cool feature! We should try to refactor the common parts but apart from that it looks good to me!

UlfNorell reviewed (changes requested):
The stack builds for older ghcs fail on Travis due to the aeson dependency. Can you check whether it's possible to support older aeson versions? I looks like `ghc-7.10.3` stack has `aeson-0.11.3.0`.

banacorn committed:
[ JSON ] Fix aeson version lower bound

banacorn committed:
[ JSON ] Fix user manual for command-line options

nad commented:
Emacs has support for parsing JSON, so one option might be to use JSON to send information to Emacs as well.

banacorn commented:
There are many refactoring opportunities.

But the next thing I want to do (which, also has a huge effect on the coming code refactoring) is to make datatypes such as `Response` and `DisplayInfo` less "forgetful".

https://github.com/agda/agda/blob/45cee704933f5fe14525e89cd6e278bc632fd958/src/full/Agda/Interaction/Response.hs#L75-L84

As you can see, information that are to be displayed by the editors have already been encoded as `String` in this stage. To recover those lost information, I had to create [dozens of parsers](https://github.com/banacorn/agda-mode/blob/master/src/parser/error.ts) in JS (which is not so fun).

I want to try and see how much I can do to postpone the stage of serialization, and refactor these modules accordingly.

banacorn committed:
[ JSON ] checkOpts: one guard to rule them all

As @gallais suggests that it does not make sense to use input-file,
--interaction or --interaction-json together.

banacorn committed:
[ JSON ] Agda.Interaction.Highlighting.Common

Refactoring common codes out from `Agda.Interaction.Highlighting.Emacs`
and `Agda.Interaction.Highlighting.JSON`.

banacorn committed:
[ JSON ] Agda.Interaction.AgdaTop

Refactoring `Agda.Interaction.JSONTop` and `Agda.Interaction.EmacsTop`
as suggested by @gallais.

gallais commented on src/full/Agda/Interaction/Highlighting/Common.hs:
```diff
@@ -0,0 +1,57 @@
+
+-- | Creates a temporary file, writes some stuff, and returns the filepath
+writeToTempFile :: String -> IO FilePath
+writeToTempFile content = do
```
This should probably go in the `Agda/Utils` directory to make
it easier for others to notice the function and reuse it.

Given that `openTempFile` is in `System.IO`, `Agda.Utils.IO`
could be a good candidate. Or maybe a new `Agda.Utils.TempFile`?

banacorn replied:
I think `Agda.Utils.TempFile` is a good idea.

gallais commented on src/full/Agda/Interaction/Highlighting/Common.hs:
```diff
@@ -0,0 +1,57 @@
+    UTF8.hPutStr handle content
+    return filepath
+  return filepath
+
```
Style nitpick:

```haskell
filepath <- expr
return filepath
```

is equivalent to `expr`

banacorn committed:
[ JSON ] Update flag-related golden files in tests

banacorn committed:
[ JSON ] Encode Info_Version in a meaningful way

banacorn committed:
[ JSON ] Move writeToTempFile to Agda.Utils.IO.TempFile

nad commented:
> Emacs has support for parsing JSON, so one option might be to use JSON to send information to Emacs as well.

This wasn't implemented, right?
